decider-bcl2fastq/src/main/java/ca/on/oicr/pde/deciders/Bcl2fastqDecider.java

```diff
@@ -706,4 +793,86 @@ private Path writeWorkflowRunIniPropertiesToFile(WorkflowRun wr) {
             throw new RuntimeException(ex);
         }
     }
+
+    private List<ProvenanceWithProvider<SampleProvenance>> applyBasesMask(List<ProvenanceWithProvider<SampleProvenance>> samples, BasesMask basesMask) throws DataMismatchException {
+        List<ProvenanceWithProvider<SampleProvenance>> samplesWithCustomBarcodes = new ArrayList<>();
+        for (ProvenanceWithProvider<SampleProvenance> pwp : samples) {
+            SampleProvenance sp = pwp.getProvenance();
+            Barcode newBarcode = BarcodeAndBasesMask.applyBasesMask(Barcode.fromString(sp.getIusTag()), basesMask);
+            SampleProvenanceWithCustomBarcode sampleProvenanceWithCustomBarcode = new SampleProvenanceWithCustomBarcode(sp, newBarcode.toString());
+            ProvenanceWithProvider<SampleProvenance> newPwp = new ProvenanceWithProvider(pwp.getProvider(), sampleProvenanceWithCustomBarcode);
+            samplesWithCustomBarcodes.add(newPwp);
+        }
+        return samplesWithCustomBarcodes;
+    }
+
+    private void addInvalidLane(String message, String... args) {
+        String error = MessageFormat.format(message, (Object[]) args);
+        laneErrors.add(error);
+    }
+
+    private List<WorkflowRunV2> generateWorkflowRunsForLane(ProvenanceWithProvider<LaneProvenance> lp, Map<String, List<ProvenanceWithProvider<SampleProvenance>>> groupedSamples, Bcl2FastqHandler handler, boolean doDemultiplexing) throws InvalidLaneException {
+        List<String> laneErrors = new ArrayList<>();
+        List<WorkflowRunV2> workflowRuns = new ArrayList<>();
+        for (Entry<String, List<ProvenanceWithProvider<SampleProvenance>>> e : groupedSamples.entrySet()) {
+            String group = e.getKey();
+            log.info("Generating workflow run for group = {}", group);
+
+            List<ProvenanceWithProvider<SampleProvenance>> samplesForGroup = e.getValue();
+
+            final List<String> barcodeErrors = new ArrayList<>();
+            List<Barcode> workflowRunBarcodes = samplesForGroup.stream().map(ProvenanceWithProvider::getProvenance)
+                    .map(SampleProvenance::getIusTag)
+                    .map(barcodeString -> {
+                        try {
+                            return Barcode.fromString(barcodeString);
+                        } catch (DataMismatchException ex) {
+                            barcodeErrors.add(ex.toString());
+                            return null;
+                        }
+                    })
+                    .collect(Collectors.toList());
+            if (!barcodeErrors.isEmpty()) {
+                laneErrors.add(MessageFormat.format("Error while parsing barcodes for group = [{0}]:\n{1}", barcodeErrors.toString()));
```

Comment:
![BLOCKER](https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER') Not enough arguments. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2275)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') No need to call toString "method()" as formatting and string conversion is done by the Formatter. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3457)